swarm/src/connection.rs

```diff
@@ -161,27 +251,37 @@ where
                 }
             }
 
-            if !self.open_info.is_empty() {
+            if !self.pending_dial_upgrades.is_empty() {
                 match self.muxing.poll_outbound_unpin(cx)? {
                     Poll::Pending => {}
                     Poll::Ready(substream) => {
-                        let user_data = self
-                            .open_info
+                        let protocol = self
+                            .pending_dial_upgrades
                             .pop_front()
                             .expect("`open_info` is not empty");
-                        let endpoint = SubstreamEndpoint::Dialer(user_data);
-                        self.handler.inject_substream(substream, endpoint);
+
+                        self.negotiating_out.push(SubstreamUpgrade::new_outbound(
+                            substream,
+                            protocol,
+                            self.substream_upgrade_protocol_override,
+                        ));
+
                         continue; // Go back to the top, handler can potentially make progress again.
                     }
                 }
             }
 
-            match self.muxing.poll_inbound_unpin(cx)? {
-                Poll::Pending => {}
-                Poll::Ready(substream) => {
-                    self.handler
-                        .inject_substream(substream, SubstreamEndpoint::Listener);
-                    continue; // Go back to the top, handler can potentially make progress again.
+            if self.negotiating_in.len() < self.max_negotiating_inbound_streams {
+                match self.muxing.poll_inbound_unpin(cx)? {
+                    Poll::Pending => {}
+                    Poll::Ready(substream) => {
+                        let protocol = self.handler.listen_protocol();
+
+                        self.negotiating_in
+                            .push(SubstreamUpgrade::new_inbound(substream, protocol));
+
+                        continue; // Go back to the top, handler can potentially make progress again.
+                    }
                 }
             }
```

Comment:
I am afraid this leads to a deadlock with the current Yamux implementation.

`<Yamux as StreamMuxer>::poll` currently only returns `Poll::Pending`, but does not drive the underlying state machine.

https://github.com/libp2p/rust-libp2p/blob/b8c3b282eee5d6387a9cf2f64af0253e764c1172/muxers/yamux/src/lib.rs#L126-L131

The only `StreamMuxer` method driving the Yamux state machine is `<Yamux as StreamMuxer>::poll_inbound`:

https://github.com/libp2p/rust-libp2p/blob/b8c3b282eee5d6387a9cf2f64af0253e764c1172/muxers/yamux/src/lib.rs#L104-L115

Note that the Yamux state machine is the component reading and writing from/to the underlying I/O resource.

Thus, once the `self.max_negotiating_inbound_streams` limit is hit, established streams and pending outbound streams will no longer make progress.

---

**Suggestion**

Yamux does not support back-pressure on stream creation at the protocol level. Thus, the only thing a node can do is drop inbound streams. The question is where to drop inbound streams.

Ideally I would like to drop streams within https://github.com/libp2p/rust-yamux once a certain limit is hit. We could as well buffer streams in `libp2p-yamux::Yamux` and drop streams exceeding the limit there.

Reply:
> Ideally I would like to drop streams within [libp2p/rust-yamux](https://github.com/libp2p/rust-yamux) once a certain limit is hit. We could as well buffer streams in `libp2p-yamux::Yamux` and drop streams exceeding the limit there.

We don't know about the limit in the yamux multiplexer unfortunately and it would be rather convoluted to pass the actual limit in. What do you think would an appropriate limit be that we could hardcode?

Reply:
Actually, we don't need a separate limit. We can move the general `poll` to the bottom of the `poll` loop in `Connection`. That way, we can simply drop all streams that we get in `StreamMuxer::poll`. If we get there and still receive streams, it means we have hit the limit on how many inbound streams we can process.

Does this make sense?

Reply:
However, this doesn't follow the convention of "prefer local over remote work".

Reply:
> However, this doesn't follow the convention of "prefer local over remote work".

This makes me prefer the separate limit again. I think we can set this to something rather high because it should not be hit in normal operation.

Reply:
Let me know what you think of [6b98c50](https://github.com/libp2p/rust-libp2p/pull/2861/commits/6b98c50209b59957e5b85413bc2397c2bdb83f5d).

Happy to adjust the limit!